packages/asset-swapper/src/utils/aggregate.ts

```diff
@@ -0,0 +1,780 @@
+ * @param opts Options object.
+ * @return Improved orders.
+ */
+export async function improveMarketBuyAsync(
+    nativeOrders: OrderWithoutDomain[],
+    makerAmount: BigNumber,
+    opts?: Partial<ImproveOrdersOpts>,
+): Promise<OrderWithoutDomain[]> {
+    if (nativeOrders.length === 0) {
+        throw new Error(AggregationError.EmptyOrders);
+    }
+    const _opts = {
+        ...IMPROVE_ORDERS_OPTS_DEFAULTS,
+        ...opts,
+    };
+    const [orderInfos, dexQuotes] = await queryNetworkAsync(
+        nativeOrders,
+        getSampleAmounts(makerAmount, _opts.numSamples),
+        difference(BUY_SOURCES, _opts.excludedSources),
+        _opts.provider,
+        true,
+    );
+    const nativePath = pruneDustFillsFromNativePath(
+        createBuyPathFromNativeOrders(nativeOrders, orderInfos),
+        makerAmount,
+        _opts.dustThreshold,
+    );
+    const dexPaths = createPathsFromDexQuotes(dexQuotes, _opts.noConflicts);
+    const allPaths = [...dexPaths];
+    const allFills = flattenDexPaths(dexPaths);
+    if (!(_opts.excludedSources as ERC20BridgeSource[]).includes(ERC20BridgeSource.Native)) {
+        allPaths.splice(0, 0, nativePath);
+        allFills.splice(0, 0, ...nativePath);
+    }
+    const optimizer = new FillsOptimizer(_opts.runLimit, true);
+    const optimalPath = optimizer.optimize(
+        sortFillsByPrice(allFills),
+        makerAmount,
+        pickOptimalPath(allPaths, makerAmount, true),
+    );
+    if (!optimalPath) {
+        throw new Error(AggregationError.NoOptimalPath);
+    }
+    const [inputToken, outputToken] = getOrderTokens(nativeOrders[0]);
+    return createBuyOrdersFromPath(inputToken, outputToken, simplifyPath(optimalPath), _opts.bridgeSlippage);
+}
+
+/**
+ * Compute the total output for a fill path, optionally clipping the input
+ * to `maxInput`.
+ */
+export function getPathOutput(path: Fill[], maxInput?: BigNumber): BigNumber {
+    let currentInput = ZERO;
+    let currentOutput = ZERO;
+    for (const fill of path) {
+        if (maxInput && currentInput.plus(fill.input).gte(maxInput)) {
+            const partialInput = maxInput.minus(currentInput);
+            currentOutput = currentOutput.plus(getPartialFillOutput(fill, partialInput));
+            currentInput = partialInput;
+        } else {
+            currentInput = currentInput.plus(fill.input);
+            currentOutput = currentOutput.plus(fill.output);
+        }
+    }
+    return currentOutput;
+}
+
+/**
+ * Query the sampler contract to get native order statuses and DEX quotes.
+ */
+export async function queryNetworkAsync(
+    nativeOrders: OrderWithoutDomain[],
+    sampleAmounts: BigNumber[],
+    sources: ERC20BridgeSource[],
+    provider?: SupportedProvider,
+    isBuy?: boolean,
+): Promise<[OrderInfo[], DexSample[][]]> {
+    if (!provider) {
+        throw new Error(AggregationError.MissingProvider);
+    }
+    const sampler = new ERC20BridgeSamplerContract(SAMPLER_ADDRESS, provider);
```

Comment:
I'd even consider just passing in the entire `ERC20BridgeSampler`, won't have to deal with optional providers here or in `ImproveOrdersOpts`